example-utils/integration-test-utils/src/up_client_foo.rs

```diff
@@ -0,0 +1,413 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use async_broadcast::{Receiver, Sender};
+use async_std::sync::Mutex;
+use async_std::task;
+use async_trait::async_trait;
+use log::{debug, error};
+use std::collections::{HashMap, HashSet};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use up_rust::{
+    ComparableListener, UAuthority, UCode, UListener, UMessage, UMessageType, UStatus, UTransport,
+    UUri,
+};
+
+pub struct UPClientFoo {
+    #[allow(dead_code)]
+    name: Arc<String>,
+    #[allow(dead_code)]
+    protocol_receiver: Receiver<Result<UMessage, UStatus>>,
+    protocol_sender: Sender<Result<UMessage, UStatus>>,
+    listeners: Arc<Mutex<HashMap<UUri, HashSet<ComparableListener>>>>,
+    authority_listeners: Arc<Mutex<HashMap<UAuthority, HashSet<ComparableListener>>>>,
+    pub times_received: Arc<AtomicU64>,
+}
+
+impl UPClientFoo {
+    pub async fn new(
+        name: &str,
+        protocol_receiver: Receiver<Result<UMessage, UStatus>>,
+        protocol_sender: Sender<Result<UMessage, UStatus>>,
+    ) -> Self {
+        let name = Arc::new(name.to_string());
+        let listeners: Arc<Mutex<HashMap<UUri, HashSet<ComparableListener>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+        let authority_listeners: Arc<Mutex<HashMap<UAuthority, HashSet<ComparableListener>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+
+        let name_clone = name.clone();
+        let authority_listeners_clone = authority_listeners.clone();
+        let listeners_clone = listeners.clone();
+        let protocol_receiver_clone = protocol_receiver.clone();
+
+        let times_received = Arc::new(AtomicU64::new(0));
+        let times_received_task = times_received.clone();
+        task::spawn(async move {
+            let name_clone = name_clone.clone();
+            let mut protocol_receiver_clone = protocol_receiver_clone.clone();
+            let listeners_clone = listeners_clone.clone();
+
+            while let Ok(received) = protocol_receiver_clone.recv().await {
+                match &received {
+                    Ok(msg) => {
+                        let UMessage { attributes, .. } = &msg;
+                        let Some(attr) = attributes.as_ref() else {
+                            debug!("{}: No UAttributes!", &name_clone);
+                            continue;
+                        };
+
+                        match attr.type_.enum_value() {
+                            Ok(enum_value) => match enum_value {
+                                UMessageType::UMESSAGE_TYPE_UNSPECIFIED => {
+                                    debug!("{}: Type unspecified! Fail!", &name_clone);
+                                }
+                                UMessageType::UMESSAGE_TYPE_NOTIFICATION => {
+                                    let sink_uuri = attr.sink.as_ref();
```

Comment:
The content of each one of these match arms is _essentially_ the same, modulo some logging differences.

Should probably be refactored.

Reply:
Fixed this!